### PR TITLE
Performance updates and fixes for OSX DataView

### PR DESCRIPTION
**Pull request by afalkenhahn**
This is an updated version of #2352. It contains the following changes:

1. We don't use `sizeLastColumnToFit` any longer because it always causes the view to refresh which will make our code very slow.
2. For autosizing columns, the text in the last column is no longer ellipsized in case it is wider than the available space. Instead, scrollbars will be added.

### CODE REVIEW

**Review by vslavik — APPROVED** [review]
I think the changes are good.

Performance-wise, they fix the problems. wxDVC is still a little slower, but a *little* seems acceptable.

File: src/osx/dataview_osx.cpp [hunk 1]
```diff
@@ -233,76 +233,65 @@ void wxOSXDataViewModelNotifier::Resort()
 
 void wxOSXDataViewModelNotifier::AdjustRowHeight(wxDataViewItem const& item)
 {
-  if ((m_DataViewCtrlPtr->GetWindowStyle() & wxDV_VARIABLE_LINE_HEIGHT) != 0)
+  if (!m_DataViewCtrlPtr->IsFrozen() ||  ((m_DataViewCtrlPtr->GetWindowStyle() & wxDV_VARIABLE_LINE_HEIGHT) != 0))
```
Comment (vadz) [inline src/osx/dataview_osx.cpp:236]: Is this condition correct, i.e. shouldn't it be an `&&`? IOW do we really need to do anything for a frozen control here?

File: src/osx/dataview_osx.cpp [hunk 2]
```diff
@@ -233,76 +233,65 @@ void wxOSXDataViewModelNotifier::Resort()
...
   {
-      wxDataViewModel *model = GetOwner();
+    wxDataViewModel *model = GetOwner();
+
+    int height = ((m_DataViewCtrlPtr->GetWindowStyle() & wxDV_VARIABLE_LINE_HEIGHT) != 0)
```
Comment (vadz) [inline src/osx/dataview_osx.cpp:240]:
```suggestion
    int height = m_DataViewCtrlPtr->HasFlag(wxDV_VARIABLE_LINE_HEIGHT)
```
(and elsewhere)

File: src/osx/dataview_osx.cpp [hunk 3]
```diff
@@ -233,76 +233,65 @@ void wxOSXDataViewModelNotifier::Resort()
...
+        ? 20
+        : m_DataViewCtrlPtr->GetDataViewPeer()->GetRowHeight();
+    int origHeight = height;
```
Comment (vadz) [inline src/osx/dataview_osx.cpp:243]: Sorry, I don't understand what is this for, could you please explain? Where/how does `origHeight` become different from `height`? It looks like neither of them ever changes after the initial assignment, there used to be an assignment to `height` before but it has disappeared now...

File: include/wx/osx/cocoa/dataview.h [hunk 4]
```diff
@@ -501,7 +501,7 @@ class wxCocoaDataViewControl : public wxWidgetCocoaImpl,
     virtual wxDataViewColumn* GetColumn(unsigned int pos) const wxOVERRIDE;
     virtual int GetColumnPosition(wxDataViewColumn const* columnPtr) const wxOVERRIDE;
     virtual bool InsertColumn(unsigned int pos, wxDataViewColumn* columnPtr) wxOVERRIDE;
-    virtual void FitColumnWidthToContent(unsigned int pos) wxOVERRIDE;
+    virtual void FitColumnWidthToContent(unsigned int pos, bool fitRowHeight = false) wxOVERRIDE;
```
Comment (vadz) [inline include/wx/osx/cocoa/dataview.h:504]: As usual, I think the code would be much more readable if we used an enum rather than bool for this parameter, i.e. `int fitFlags = 0` with the possibility to pass `Fit_RowHeight`.

Although looking at the code this really could be called `Force_AllRows` too because it seems that all passing `true` here does is to set `calculateAllRows` to `true` in the code.

File: src/osx/cocoa/dataview.mm [hunk 5]
```diff
@@ -2320,64 +2324,71 @@ void UpdateWithRow(int row)
     // visible miscalculations, we also include all currently visible items
     // no matter what.  Finally, the value of N is determined dynamically by
     // measuring how much time we spent on the determining item widths so far.
-
+    if (calculateAllRows)
```
Comment (vadz) [inline src/osx/cocoa/dataview.mm:2327]: Is the condition here correct? It looks like we're using the optimized version for "all rows" case, when I'd expect it to iterate over all rows. Or maybe the variable is named incorrectly instead?

File: src/osx/cocoa/dataview.mm [hunk 6]
```diff
@@ -2305,6 +2305,10 @@ void UpdateWithRow(int row)
 
     MaxWidthCalculator calculator(m_OutlineView, column, pos);
 
+    bool calculateAllRows = fitRowHeight || ((GetColumn(pos)->GetWidthVariable() == wxCOL_WIDTH_AUTOSIZE)
+        || (m_expanderWidth == -1 && column == [m_OutlineView outlineTableColumn]));
```
Comment (vadz) [inline src/osx/cocoa/dataview.mm:2309]: I'd really appreciate a comment explaining what's going on here and what does this variable mean because I think I'm misunderstanding its purpose.

File: src/osx/cocoa/dataview.mm [hunk 7]
```diff
@@ -2389,17 +2400,34 @@ void UpdateWithRow(int row)
...
 
-        [m_OutlineView sizeLastColumnToFit];
+        // Compute the remaining space for the last column; note that we
+        // deliberately don't use sizeLastColumnToFit() because it causes a
+        // refresh of the NSTableView; using sizeLastColumnToFit() would also
+        // require us to temporarily enable NSTableColumnAutoresizingMask on
+        // the NSTableColumn because it doesn't seem to work without it
+        NSSize size = [[m_OutlineView superview] bounds].size;
+        size.width -= (noOfColumns * [m_OutlineView intercellSpacing].width);
+        for ( UInt32 k = 0; k < noOfColumns - 1; k++ ) {
```
Comment (vadz) [inline src/osx/cocoa/dataview.mm:2410]:
```suggestion
        for ( UInt32 k = 0; k < noOfColumns - 1; k++ )
        {
```

File: src/osx/cocoa/dataview.mm [hunk 8]
```diff
@@ -2389,17 +2400,34 @@ void UpdateWithRow(int row)
...
+            NSTableColumn *c = GetColumn(k)->GetNativeData()->GetNativeColumnPtr();
+            size.width -= [c width];
+        }
+        
+        int lastColWidth = ceil(size.width);
+           
+        if ( GetColumn(pos)->GetWidthVariable() == wxCOL_WIDTH_AUTOSIZE && autoWidth > lastColWidth ) lastColWidth = autoWidth;
```
Comment (vadz) [inline src/osx/cocoa/dataview.mm:2417]:
```suggestion
        if ( GetColumn(pos)->GetWidthVariable() == wxCOL_WIDTH_AUTOSIZE && autoWidth > lastColWidth )
            lastColWidth = autoWidth;
```

File: src/osx/cocoa/dataview.mm [hunk 9]
```diff
@@ -2733,6 +2761,9 @@ void UpdateWithRow(int row)
 {
     [m_OutlineView setRowHeight:wxMax(height, GetDefaultRowHeight())];
 }
+int wxCocoaDataViewControl::GetRowHeight() {
```
Comment (vadz) [inline src/osx/cocoa/dataview.mm:2764]:
```suggestion

int wxCocoaDataViewControl::GetRowHeight()
{
```

File: src/osx/cocoa/dataview.mm [hunk 10]
```diff
@@ -2753,7 +2784,10 @@ void UpdateWithRow(int row)
 
 void wxCocoaDataViewControl::OnSize()
 {
-    [m_OutlineView sizeLastColumnToFit];
+    UInt32 const noOfColumns = [[m_OutlineView tableColumns] count];
+    if (noOfColumns) {
```
Comment (vadz) [inline src/osx/cocoa/dataview.mm:2788]:
```suggestion
    if (noOfColumns)
    {
```

**Review by vadz — COMMENTED** [review]
Sorry, I don't really understand the logic in `FitColumnWidthToContent()`, could you please explain it?

You can ignore the minor style problems and even the `bool` one as I could just fix those when merging myself but it would be really nice to understand what's going on here. TIA!

File: src/osx/cocoa/dataview.mm (hunk 5 above)
Comment (vslavik) [reply src/osx/cocoa/dataview.mm:2327]: Trying to explain on behalf of @dkulp, so please correct me if I got something wrong…

These changes are trying to fix previously merged changes to this code (that IMHO in retrospect shouldn't have been merged). Part of the problem with them was that they changed `FitColumnWidthToContent()` to do much more than the name says —  on its face, it still does column width fitting, but *also* calculates (max) row height, and column expander width. `FitColumnWidthToContent` is no longer a fitting name. `MaxWidthCalculator` is now misnamed too. And the function is, because of that change of purpose, called much more often than it used to.

This is a problem, because it kills performance. So @dkulp is trying to limit pointless calculations. One of the optimizations is to avoid iterating over all the rows if we know it won't provide any new information.

So yes, it is a badly named variable. It should be e.g. `shouldCalculateRows`.

File: include/wx/osx/cocoa/dataview.h (hunk 4 above)
Comment (vslavik) [reply include/wx/osx/cocoa/dataview.h:504]: > Although looking at the code this really could be called Force_AllRows too

`Fit_RowHeight` is semantic and says "we need to update height calculation when called from this place". That doing that forces iteration over all rows is an implementation detail that may change.

OTOH, it only is used in one place (Thaw) where "force all calculations" *is* appropriate meaning.

File: src/osx/dataview_osx.cpp (hunk 1 above)
Comment (vslavik) [reply src/osx/dataview_osx.cpp:236]: You're right, I think.

File: src/osx/dataview_osx.cpp (hunk 3 above)
Comment (vslavik) [reply src/osx/dataview_osx.cpp:243]: A little lower, there's `height = wxMax( height, renderer->GetSize().y );` within the loop iterating over columns. Later, it is used to avoid explicit call to `GetDataViewPeer()->SetRowHeight()` if the height didn't change.

**Comment by discnl** [comment]
Regarding automatic row height changes which were also discussed in PR #2370, the changes of d2fc88c03dc for macOS were committed based on the assumption that wxGTK already behaved the same: for a DVC without `wxDV_VARIABLE_LINE_HEIGHT` row height is automatically determined by the highest row (e.g. a row with a bigger icon). This isn't actually the case with GTK3 from what I see now, tested with Mint 20. By default wxGTK appears to behave as if `wxDV_VARIABLE_LINE_HEIGHT` was specified, and only when explicitly using `SetRowHeight()` a uniform row height is used. See also this old [comment](http://trac.wxwidgets.org/ticket/12749#comment:6) and current [gtk_wx_cell_renderer_get_size](https://github.com/wxWidgets/wxWidgets/blob/b52f00492e4c6ff199488e1997fff19a8e158bb7/src/gtk/dataview.cpp#L1555) code.
So with wxGTK even if a taller row gets exposed via expansion it is only that row which then changes height and other displayed rows maintain the same height. The wxGTK behaviour can be demonstrated by using icons with a different height, see this [diff](https://github.com/discnl/wxWidgets/commit/53e29909a4c531304ce8e86c8b91263366fcd5ed.patch) to the dataview sample. The first DVC page has a row with a taller icon hidden under its collapsed node, while the second page uses taller icons for odd rows.

With the modified sample and wxMSW there's no row height adjustments at all as expected, ticket [#18868](http://trac.wxwidgets.org/ticket/18868) remained open for supporting that on Windows. Rows keep their initial height and taller icons are truncated.

Because formerly there was actually no platform making use of automatic uniform row height I'm hesitant to add something [similar](http://trac.wxwidgets.org/attachment/ticket/18868/msw-dvc-image-rowheight.patch) as d2fc88c03dc to the generic version of DVC. Also because a user could already call `SetRowHeight()` to have uniform heights.

**Comment by vadz** [comment]
Sorry, I still didn't have time to get back to this, so, just to be clear: am I right that both of you (@vslavik and @discnl) believe that everything here is correct (modulo cosmetic/style changes) and that this should be merged?

**Comment by vslavik** [comment]
Sorry @vadz for my slowness. I don't think this is mergeable anymore.

1. I have new issues with 57e2efa (it's broken on macOS 11 and makes the last column too large even when it shouldn't).

2. More importantly, if I understand @discnl correctly, the commits that broke wxDVC performance ( 36ea7ff4d61e174a2d116819f98e66f22a7f5b8a, 7555d1b2459949c128d6ee3a473669a1a134f124) were merged under incorrect understanding of what the behavior should be.

The changes intended to make wxOSX behavior consistent with other platforms, but actually *no* platform natively has this strange behavior where all rows are autosized to have uniform height matching the row with the largest height. And it isn't implemented in our generic wxDVC implementation on Windows. In other words, this behavior was newly introduced for wxOSX.

This seems reasonable to me. Autosizing row heights to accommodate unknown icon sizes dynamically did not make sense to me and I *was* surprised that some platforms would do it (but didn't question it when that argument was used).

In light of @discnl's testing, my vote would certainly be to revert the offending changes instead and don't merge this PR (it does make sense to cherry-pick @discnl's freeze/thaw optimization, though). A combination of `wxDV_VARIABLE_LINE_HEIGHT` (which isn't implemented in wxOSX, but contrary to a comment in the code seems to be implement*able*) and explicit `SetRowHeight()` to enlarge the rows is sufficient for reasonable UIs.

If we don't need the dynamically-resize-uniform-row-height behavior for compatibility/consistency, I struggle to think of a reason why we'd want to introduce it. It is not natively supported by OS controls and is inherently  ~~evil~~ O(n). That's not a burden wxDVC should carry, IMNSHO, or at least not without it being explicitly enabled with a dedicated flag.

**Comment by vadz** [comment]
@vslavik: At this point I'd be glad if you just did whatever you consider to be the best, even if it just consists in reverting the old changes. Not fully fixing the original problem would be unfortunate, of course, but I just can't justify spending tons more of time on it when there are much more important things to do before 3.2, so any solution not resulting in bad regressions would be fine to me for now. Could you please do it? TIA!

And IMO we need to rethink the process for future changes from @softwarefailure and @dkulp, because as much as I appreciate your help with the different Mac issues, the current way of submitting many small partially working patches and then recombining them clearly doesn't work well at all.

**Comment by softwarefailure** [comment]
> The changes intended to make wxOSX behavior consistent with other platforms, but actually no platform natively has this strange behavior where all rows are autosized to have uniform height matching the row with the largest height. And it isn't implemented in our generic wxDVC implementation on Windows. 

It definitely is. See those two screenshots, as soon as the node that contains an image higher than the row height is unfolded, the global row height setting is adapted to the size of that row.

folded
unfolded

And I'm pretty sure it behaves the same way on GTK. 

Anyway, I'd like to point out that the row height thing is not really what this PR is all about. Keep in mind that the original issue addressed by my patches was [#14939](https://trac.wxwidgets.org/ticket/14939). So if you revert anything here make sure not to re-introduce the issue of #14939 again because that was really a major issue and limitation in wxDVC on Cocoa. And most of my patches actually address this issue. So if you really don't want the automatic row height adaptation, I'd suggest to just disable that particular feature and try to keep as much of the rest as possible. I don't have a macOS 11 system here right now, so @vslavik should probably debug the issues he's seeing on macOS 11 himself. They probably can be fixed quite easily.

TLDR; I don't recommend to ditch everything in this PR as it contains several important fixes. Instead, I recommend to just disable the automatic row height adaptation if you don't want it and use the rest of the PR with fixes for macOS 11 applied.

**Comment by discnl** [comment]
> Sorry, I still didn't have time to get back to this, so, just to be clear: am I right that both of you (@vslavik and @discnl) believe that everything here is correct (modulo cosmetic/style changes) and that this should be merged?

I hadn't tried this PR before but noticed performance degradation was mentioned at several places, involving changes for auto-uniform row height as well as at least (last/auto) colum widths. Since wxGTK didn't turn out to have auto-uniform row height behaviour I wanted to raise that issue because wxOSX was changed with the assumption it would be consistent with wxGTK (only, there was no mention of wxMSW behaving like that at the time).

FWIW tested with 10.13 the degradation performance appears gone and it feels on par with 3.1.4 when toggling checkboxes in the MyListModel page of the dataview sample. The last column behaves better, making it wider instead of narrower when resizing another column. That's similar to at least wxMSW, the only difference (just an observation currently) is that with wxMSW the last column first shrinks to its minimal width (so correctly showing no horizontal scrollbar yet) before growing.

With macOS 11 I'm seeing the same as Václav: the last column is too wide and you can scroll horizontally (apparently visualized by not showing a highlight with rounded corners if the row doesn't fit).



> Sorry @vadz for my slowness. I don't think this is mergeable anymore.
> 
>     2. More importantly, if I understand @discnl correctly, the commits that broke wxDVC performance ( [36ea7ff](https://github.com/wxWidgets/wxWidgets/commit/36ea7ff4d61e174a2d116819f98e66f22a7f5b8a), [7555d1b](https://github.com/wxWidgets/wxWidgets/commit/7555d1b2459949c128d6ee3a473669a1a134f124)) were merged under incorrect understanding of what the behavior should be.

I'm not familiar with those but based on https://github.com/wxWidgets/wxWidgets/commit/36ea7ff4d61e174a2d116819f98e66f22a7f5b8a#r50173486 that part of commit 36ea7ff is only needed for auto-uniform row height reasons. 7555d1b seems to be about last column issues and not directly related to auto-uniform changes. It may be related to the mentioned performance degration if `FitColumnWidthToContent` is involved of course though. It seems a bit tricky to me to distinguish between the changes (and those for `wxCOL_WIDTH_AUTOSIZE`) between the various commits.

> In light of @discnl's testing, my vote would certainly be to revert the offending changes instead and don't merge this PR (it does make sense to cherry-pick @discnl's freeze/thaw optimization, though).

They are @dkulp changes, but indeed the freeze/thawing seems fine and stand-alone.

> A combination of `wxDV_VARIABLE_LINE_HEIGHT` (which isn't implemented in wxOSX, but contrary to a comment in the code seems to be implement_able_) and explicit `SetRowHeight()` to enlarge the rows is sufficient for reasonable UIs.

Agreed, and if a user wants uniform row height to be changed as a result of expanding a node with taller row then that's still possible.

I wonder if (exposing) a calculated max row height (which in itself is cheap since `GetSize()` already gets called) could still be useful in some ways, e.g. for a user with determining what to pass to `SetRowHeight()`. If those `MaxWidthCalculator` changes stay then it indeed would need renaming as [mentioned](https://github.com/wxWidgets/wxWidgets/pull/2381#discussion_r643096369), which wasn't done yet because similar changes were, formerly, planned for generic wxDVC and then it could be merged with `include/wx/generic/private/widthcalc.h` .


> > The changes intended to make wxOSX behavior consistent with other platforms, but actually no platform natively has this strange behavior where all rows are autosized to have uniform height matching the row with the largest height. And it isn't implemented in our generic wxDVC implementation on Windows.
> 
> It definitely is. See those two screenshots, as soon as the node that contains an image higher than the row height is unfolded, the global row height setting is adapted to the size of that row.

@softwarefailure How to reproduce this with the dataview sample? I tried with [this](https://github.com/discnl/wxWidgets/commit/53e29909a4c531304ce8e86c8b91263366fcd5ed.patch) under Win 10 and XP but am unable to get the imaged behaviour, with these changes the row height unfortunately stayed the same and the image is truncated (as [expected](http://trac.wxwidgets.org/ticket/18868#comment:12)).

> And I'm pretty sure it behaves the same way on GTK.

What I am seeing with wxGTK is that, without calling `SetRowHeight()` to get uniform row heights, it works as if `wxDV_VARIABLE_LINE_HEIGHT` was specified, [more info](https://github.com/wxWidgets/wxWidgets/pull/2381#issuecomment-860253897). So if all rows have equal height (e.g. because they all use the same icon) then it appears as if a uniform row height is used, but each row just uses its own calculated height.

File: src/osx/dataview_osx.cpp (hunk 1 above)
Comment (discnl) [reply src/osx/dataview_osx.cpp:236]: Usage of `||` here results in the adjustments being delayed only if the view is frozen and not using variable row height, which corresponds with the commit message:
> If the view is Frozen and it's not using variable row heights, delay [...]

So with `wxDV_VARIABLE_LINE_HEIGHT` it behaves as before, I suppose either because it's still required or was only tested without `wxDV_VARIABLE_LINE_HEIGHT`.

**Comment by softwarefailure** [comment]
> @softwarefailure How to reproduce this with the dataview sample? I tried with [this](https://github.com/discnl/wxWidgets/commit/53e29909a4c531304ce8e86c8b91263366fcd5ed.patch) under Win 10 and XP but am unable to get the imaged behaviour, with these changes the row height unfortunately stayed the same and the image is truncated (as [expected](http://trac.wxwidgets.org/ticket/18868#comment:12)).

You need to use `wxCOL_WIDTH_AUTOSIZE` for this to work. The code responsible for adapting the global row height this is in `wxDataViewCtrl::GetBestColumnWidth()` in the generic control:

```
    if (rowHeight > curHeight)
    {
          const_cast<wxDataViewCtrl*>(this)->SetRowHeight(rowHeight);
    }   
```

This will automatically change the global row height as soon as rows higher than the current height are unfolded if `wxCOL_WIDTH_AUTOSIZE` is active.

Also, I've now set up a macOS 11 system and tried to reproduce the problems reported by @vslavik and @discnl but I don't really understand the problem. Could anybody describe what the problem on macOS 11 is with my patch and how I can reproduce it? Maybe attach a screenshot as well. I might be able to come up with a fix then.

**Comment by discnl** [comment]
> You need to use `wxCOL_WIDTH_AUTOSIZE` for this to work.

Then it indeed works but only with [such](http://trac.wxwidgets.org/attachment/ticket/18868/msw-dvc-image-rowheight.patch) changes, the quoted code is definitely not in [master](https://github.com/wxWidgets/wxWidgets/blob/2867aa977878ef05d34f9c6e941cbf4cdd15cac7/src/generic/datavgen.cpp#L6018-L6048). Sorry for not mentioning earlier whether local changes could be involved with the wxMSW behaviour, I did try with patches for ticket [#18868](http://trac.wxwidgets.org/attachment/ticket/18868) but forgot the somewhat peculiar requirement for auto-uniform height to work through `wxCOL_WIDTH_AUTOSIZE`.

> Also, I've now set up a macOS 11 system and tried to reproduce the problems reported by @vslavik and @discnl but I don't really understand the problem. Could anybody describe what the problem on macOS 11 is with my patch and how I can reproduce it? Maybe attach a screenshot as well. I might be able to come up with a fix then.

With macOS 11 after startup of the dataview sample I can already scroll the DVC horizontally (by default scrollbars aren't visible until scrolling), while with at least 10.13 and 10.15 I correctly can't. With today's master there's no horizontal scrolling after startup of the sample on macOS 11.

I also checked with wxGTK 3 and it behaves the same as wxMSW when increasing width of the first column: the last column shrinks up to some minimal width, when that width is reached (and maintained) only then horizontal scrolling is introduced. This is with today's master and the initial page of the dataview sample.

**Comment by softwarefailure** [comment]
> Then it indeed works but only with [such](http://trac.wxwidgets.org/attachment/ticket/18868/msw-dvc-image-rowheight.patch) changes, the quoted code is definitely not in [master]

Oops, right. Too many good patches seem to get forever lost on wxTrac. I guess that's the reason why I'm maintaining my own custom build of wxWidgets.

> With macOS 11 after startup of the dataview sample I can already scroll the DVC horizontally (by default scrollbars aren't visible until scrolling), while with at least 10.13 and 10.15 I correctly can't. With today's master there's no horizontal scrolling after startup of the sample on macOS 11.

Can't reproduce this. This is what the dataview sample looks like with #2381 on macOS 11 here:

Bildschirmfoto 2021-06-23 um 16 23 59

As you can see, there's no horizontal scrollbar...

**Comment by discnl** [comment]
> As you can see, there's no horizontal scrollbar...

By default scrollbars aren't visible until scrolling. From the highlight in the screenshot I can see that this DVC will be able to scroll horizontally.

**Comment by vadz** [comment]
> > Then it indeed works but only with [such](http://trac.wxwidgets.org/attachment/ticket/18868/msw-dvc-image-rowheight.patch) changes, the quoted code is definitely not in [master]
> 
> Oops, right. Too many good patches seem to get forever lost on wxTrac. I guess that's the reason why I'm maintaining my own custom build of wxWidgets.

I don't know how do you manage to make such wrong conclusion from the correct premise. Yes, it's difficult to apply Trac patches because they don't apply to master cleanly and often destructively interfere with each other. No, the best solution is not to avoid submitting any patches at all, but to create pull requests that actually can be merged cleanly.

I thought this would be obvious to all involved by now, but if it isn't, I don't know what can I do to convince you about it.

**Comment by softwarefailure** [comment]
> I don't know how do you manage to make such wrong conclusion from the correct premise. Yes, it's difficult to apply Trac patches because they don't apply to master cleanly and often destructively interfere with each other. No, the best solution is not to avoid submitting any patches at all, but to create pull requests that actually can be merged cleanly.

Well, let's put the blame on @discnl because in [comment 14](http://trac.wxwidgets.org/ticket/18868#comment:14)  he promised to make a PR but apparently forgot to do so.

**Comment by vadz** [comment]
I'm not interested at all in assigning blame, only in improving the process in the future.

**Comment by vslavik** [comment]
> > As you can see, there's no horizontal scrollbar…
> 
> By default scrollbars aren't visible until scrolling. From the highlight in the screenshot I can see that this DVC will be able to scroll horizontally.

Forcing `[setSelectionHighlightStyle:NSTableViewSelectionHighlightStyleSourceList]` makes it nicely visible:

Screen Shot 2021-07-12 at 7 55 55

macOS 11 chooses style automatically in some contexts, without having to force it like above, and these styles include padding. See https://stackoverflow.com/questions/64791699/how-to-remove-nstableview-indentation-on-big-sur for an illustration. 

So the issue is fundamental: manually explicitly calculating width is prone to break with future OS releases that tweak it.

**Comment by vadz** [comment]
Does anybody have any idea what to do about this? I really don't want/can't debug all this code and fix it myself, but AFAIU this can't be applied as is and yet we have to do something in order to avoid the performance regression mentioned by @dkulp in [#19166](https://trac.wxwidgets.org/ticket/19166). This ticket blames 36ea7ff4d61e174a2d116819f98e66f22a7f5b8a for the slowdown, will just reverting this commit fix the regression (at the cost of reopening [#14939](https://trac.wxwidgets.org/ticket/14939), of course)?

TIA for any help.

**Comment by discnl** [comment]
> Does anybody have any idea what to do about this?

Not yet, I at least wanted to figure out what's been going on but untangling the code is another thing. Provided automatic uniform row height changes are not desired  then d2fc88c03dcba3f2d533030d3fa83438dac016af already could be reverted to disable that. And, IIRC, then there's some commits that are either related solely to those row height changes or mixed along with e.g. improving `wxCOL_WIDTH_AUTOSIZE` support.


>This ticket blames [36ea7ff](https://github.com/wxWidgets/wxWidgets/commit/36ea7ff4d61e174a2d116819f98e66f22a7f5b8a) for the slowdown, will just reverting this commit fix the regression (at the cost of reopening [#14939](https://trac.wxwidgets.org/ticket/14939), of course)?

(That seemed easy to quickly check but unfortunately it doesn't revert cleanly.)

**Comment by vadz** [comment]
Unfortunately we had to merge #2461 to fix the unacceptable performance regressions after the previous changes and this one can't be applied any more, so I'm closing it. Of course, it does mean that we haven't really improved anything after spending so much time on this, which is sad, but at least wxDVC is going to be usable under Mac in 3.2.0.

File: src/osx/cocoa/dataview.mm [hunk 11]
```diff
@@ -2400,39 +2400,34 @@ void UpdateWithRow(int row)
...
-            if (minW == 0) {
-                [column setMinWidth:0];
-            }
+        // Compute the remaining space for the last column; note that we
+        // deliberately don't use sizeLastColumnToFit() because it causes a
+        // refresh of the NSTableView; using sizeLastColumnToFit() would also
+        // require us to temporarily enable NSTableColumnAutoresizingMask on
+        // the NSTableColumn because it doesn't seem to work without it
+        NSSize size = [[m_OutlineView superview] bounds].size;
+        size.width -= (noOfColumns * [m_OutlineView intercellSpacing].width);
```
Comment (discnl) [inline src/osx/cocoa/dataview.mm:2409]: This seems wrong when having one column, as there is no neighbouring column requiring spacing. However if the left and right marging are treated as cell spacing then changing this to:

```suggestion
        size.width -= ((noOfColumns + 1) * [m_OutlineView intercellSpacing].width);
```

gets rid of horizontal scrollbars with macOS 11. I can increase `size.width` by 2 more, with 3 a scrollbar appears again. I haven't been able to find out more about the margin extent.

(I'm aware this PR is closed but didn't want this info to go completely missing)
